### PR TITLE
Adds Capture Functionality

### PR DESCRIPTION
**Pull request by gfs**
This adds new "Capture" functionality.  The type of captured returned depends on the query chosen, the type of the field and what was found.

The primary motivation behind this was being able to get Regex context but it works for all the operations.

You only capture data if you set the "Capture" flag on the Clause for that data.

The walkthrough has been updated with a capture example and the operations page has been updated with details about what each operation captures. 

https://github.com/microsoft/OAT/wiki/Walkthrough
https://github.com/microsoft/OAT/wiki/Operations

### CODE REVIEW

File: OAT.Tests/CaptureTests.cs [hunk 1]
```diff
@@ -0,0 +1,835 @@
...
+        public void TestEQCapture()
+        {
+            var RuleName = "Equals Capture";
+            var analyzer = new Analyzer();
+
+            var eqCaptureRule = new Rule(RuleName)
+            {
+                Clauses = new List<Clause>()
+                {
+                    new Clause(OPERATION.EQ, "StringField")
```
Comment (scovetta) [inline OAT.Tests/CaptureTests.cs:39]: Any reason why the OPERATION enum is all caps?

File: OAT.Tests/VehicleDemo.cs [hunk 2]
```diff
@@ -51,20 +51,24 @@ int GetCost(Vehicle vehicle, Analyzer analyzer, IEnumerable<Rule> rules)
             return ((VehicleRule)analyzer.Analyze(rules, vehicle).MaxBy(x => x.Severity).FirstOrDefault())?.Cost ?? 0;
         }
 
-        public (bool Applies, bool Result) OverweightOperationDelegate(Clause clause, IEnumerable<string>? valsToCheck, IEnumerable<KeyValuePair<string, string>> dictToCheck, object? state1, object? state2)
+        public (bool Applies, bool Result, ClauseCapture? Capture) OverweightOperationDelegate(Clause clause, object? state1, object? state2)
```
Comment (lb5tr) [inline OAT.Tests/VehicleDemo.cs:54]: I think this return types merits seprate struct/class.

File: OAT/Analyzer.cs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License.
 using KellermanSoftware.CompareNetObjects;
```
Comment (lb5tr) [inline OAT/Analyzer.cs:2]: this diff is massive :o

File: OAT.Tests/CaptureTests.cs [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+                            "Secret.*([Mm]agic).*"
+                        },
+                        Capture = true
+                    }
+                }
+            };
+
+            var testString = "The Secret is Magic";
+            var res = analyzer.GetCapture(regexCapture, testString, null);
+            Assert.IsTrue(res.Result?.Captures.Any(x => x is TypedClauseCapture<Match> y && y.Result.Groups[1].Value == "Magic") is true);
```
Comment (scovetta) [inline OAT.Tests/CaptureTests.cs:551]: I'm sure this is correct, but what does the `is true` at the end of this line do?

File: OAT.Tests/CaptureTests.cs (hunk 1 above)
Comment (gfs) [reply OAT.Tests/CaptureTests.cs:39]: Just stylistically chose enums to be all caps.

File: OAT.Tests/CaptureTests.cs (hunk 4 above)
Comment (gfs) [reply OAT.Tests/CaptureTests.cs:551]: The result of that evaluation is truthy (bool?) But assert doesn't work on truthy values so you need to check if it's actually a bool, so I check is true - both not null and not false.

**Comment by gfs** [comment]
Thanks for the reviews. I'll give an overview in the team meeting on Monday for additional design context.

File: OAT.Tests/CaptureTests.cs (hunk 1 above)
Comment (gfs) [reply OAT.Tests/CaptureTests.cs]: See update.  I changed the Operation enum to be PascalCase.

**Pull request merged**